yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,324 @@
+package recorder
+
+import (
+	"bytes"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"hash/fnv"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"unicode"
+
+	"go.uber.org/yarpc/transport"
+	"golang.org/x/net/context"
+	"gopkg.in/yaml.v2"
+)
+
+var recorderFlag = flag.String("recorder", "replay",
+	`replay: replay from recorded request/response pairs.
+overwrite: record all request/response pairs, overwriting records.
+append: replay existing and record new request/response pairs.`)
+
+// Recorder records & replay yarpc requests on the client side.
+//
+// For recording, the client must be connected and able to issue requests to a
+// remote service. Every request and its response is recorded into a YAML file,
+// under the directory "testdata/recordings".
+//
+// During replay, the client doesn't need to be connected, for any recorded
+// request Recorder will return the recorded response. Any new request will
+// return an error.
+type Recorder struct {
+	mode       Mode
+	logger     Logger
+	recordsDir string
+}
+
+const recorderDir = "testdata/recordings"
+
+// Mode the different replay and recording modes.
+type Mode int
+
+const (
+	// invalidMode is private and used to represent invalid modes.
+	invalidMode = iota + 1
+
+	// Replay replays stored request/response pairs.
+	Replay
+
+	// Overwrite will record on file all request/response pairs, overwriting
+	// existing records.
+	Overwrite
+
+	// Append will record on file all new request/response pairs, keeping
+	// existing record without modification.
+	Append
+)
+
+func (m Mode) toHumanString() string {
+	switch m {
+	case Replay:
+		return "replaying"
+	case Overwrite:
+		return "recording (overwrite)"
+	case Append:
+		return "recording (append)"
+	}
+	panic("Unreachable")
+}
+
+// ModeFromString converts an English string of a mode to a `Mode`.
+func ModeFromString(s string) (Mode, error) {
+	switch s {
+	case "replay":
+		return Replay, nil
+	case "overwrite":
+		return Overwrite, nil
+	case "append":
+		return Append, nil
+	}
+	return invalidMode, fmt.Errorf(`invalid mode: "%s"`, s)
+}
+
+// Logger is an interface used by the recorder for logging and reporting fatal
+// errors. It is intentionally made to match with testing.T.
+type Logger interface {
+	// Logf must behaves similarly to testing.T.Logf.
+	Logf(format string, args ...interface{})
+
+	// Fatal should behaves similarly to testing.T.Fatal. Namely, it must abort
+	// the current test.
+	Fatal(args ...interface{})
+}
+
+// NewRecorder returns a Recorder in whatever mode specified via the
+// `recorderFlag`.
+//
+// The new Recorder instance is a yarpc filter middleware. It takes a logger as
+// argument compatible with `testing.T`. It is designed to be used in testing
```

Comment:
Also, no need to be mentioned it's designed to be used for testing. It's in the yarpctest package tree.